cgp/genome.py

```diff
@@ -267,6 +267,181 @@ def randomize(self, rng: np.random.RandomState) -> None:
+
+        Gene values are updated to correspond to the new address of the corresponding node.
+
+        Invalid gene values (from unused arities of eg. constant operator nodes)
+        are replaced by (pseudo-) random picks of permissible values for that position.
+
+        Parameters
+        ----------
+        rng : numpy.RandomState
+            Random number generator instance.
+
+        Returns
+        ----------
+        None
+        """
+        if not self._n_rows == 1:
+            Warning("n_rows must be 1, reorder is not applied")
+            return
+        if not self._levels_back == self._n_columns:
+            Warning("levels_back must be equal n_columns, reorder is not applied")
+            return
+
+        dna_for_reorder = self._dna.copy()
+
+        dependencies = self._determine_dependencies()
+
+        addable_nodes = set(int(idx_str) for idx_str, item in dependencies.items() if not item)
+
+        node_idx = self._n_inputs  # First position to be placed is after inputs
+        used_idx_list: List = []
+
+        while len(addable_nodes) > 0:
+            # pick one of the addable nodes
+            current_idx = rng.choice(list(addable_nodes))
+
+            dna_for_reorder = self._exchange_dna_segments(
+                dna_for_reorder, idx_old_location=current_idx, idx_new_location=node_idx
+            )
+
+            for _, current_dependencies in dependencies.items():
+                current_dependencies.discard(current_idx)
+
+            addable_nodes = self._update_addable_nodes(
+                dependencies, used_idx_list, idx_used=current_idx
+            )
+            node_idx += 1
+
+        self._update_gene_values(dna_for_reorder, used_idx_list)
+        self._replace_invalid_gene_values(dna_for_reorder, rng)
+
+        self.dna = dna_for_reorder
+
+    def _exchange_dna_segments(
+        self, dna_new: List[int], idx_old_location: int, idx_new_location: int
+    ) -> List[int]:
+        """ Moves a nodes dna from its old node location to a new location. """
+
+        node_dna = self._dna[
+            idx_old_location
+            * self._length_per_region : (idx_old_location + 1)
+            * self._length_per_region
+        ]
+        dna_new[
+            idx_new_location
+            * self._length_per_region : (idx_new_location + 1)
+            * self._length_per_region
+        ] = node_dna
+
+        return dna_new
+
+    def _update_addable_nodes(
+        self, dependencies: Dict, used_idx_list: List[int], idx_used: int
+    ) -> set:
+        """ Update the set of addable nodes,
+        by reevaluating which nodes haves no dependencies
+        and remove nodes which were already placed from the list.
+        """
+        addable_nodes = set(int(idx_str) for idx_str, item in dependencies.items() if not item)
+        used_idx_list.append(idx_used)
+        addable_nodes = addable_nodes.difference(used_idx_list)
+
+        return addable_nodes
```

Comment:
```suggestion
        return addable_nodes, used_idx_list
```
since you modify both of these lists, i would return both of them. i'm not sure what the most pythonic way is though. @mschmidt87 any input?